internal/provider/roles/client.go

```diff
@@ -0,0 +1,195 @@
+package roles
+
+import (
+	"context"
+	"fmt"
+	"github.com/hashicorp/terraform-plugin-framework/attr"
+	"github.com/hashicorp/terraform-plugin-framework/diag"
+	"github.com/hashicorp/terraform-plugin-framework/types"
+	"github.com/hashicorp/terraform-plugin-log/tflog"
+	"github.com/permitio/permit-golang/pkg/models"
+	"github.com/permitio/permit-golang/pkg/permit"
+)
+
+type RoleClient struct {
+	client *permit.Client
+}
+
+type RoleMethods interface {
+	RoleRead(ctx context.Context, data RoleModel) (RoleModel, error)
+	RoleCreate(ctx context.Context, actions map[string]models.ActionBlockEditable, resourcePlan *RoleModel) error
+	RoleUpdate(ctx context.Context, resourcePlan *RoleModel) error
+}
+
+func (d *RoleClient) RoleRead(ctx context.Context, data RoleModel) (RoleModel, error, diag.Diagnostics) {
+	var (
+		resourceKeyOrId string
+		diags           diag.Diagnostics
+	)
+	if data.Key.IsNull() {
+		resourceKeyOrId = data.Id.ValueString()
+	} else {
+		resourceKeyOrId = data.Key.ValueString()
+	}
+	role, err := d.client.Api.Roles.Get(ctx, resourceKeyOrId)
+	if err != nil {
+		return RoleModel{}, err, diags
+	}
+	// Converting lists from data input to compare with the role read from the API
+	permissionsData := make([]string, 0)
+	for _, permission := range data.Permissions.Elements() {
+		permissionsData = append(permissionsData, RemoveDoubleQuotes(permission.String()))
+	}
+
+	extendsData := make([]string, 0)
+	for _, extend := range data.Extends.Elements() {
+		extendsData = append(extendsData, RemoveDoubleQuotes(extend.String()))
+	}
+
+	if !compareSlicesValues(role.Permissions, permissionsData) || !compareSlicesValues(role.Extends, extendsData) {
+		diags.AddError(
+			"Permissions or extends mismatch to the created role",
+			fmt.Sprintf("Permissions from data: %v, Permissions from api: %v, Extends from data: %v, Extends from api: %v ", permissionsData, role.Permissions, extendsData, role.Extends),
+		)
+	}
+	state := RoleModel{
+		Id:             types.StringValue(role.Id),
+		OrganizationId: types.StringValue(role.OrganizationId),
+		ProjectId:      types.StringValue(role.ProjectId),
+		EnvironmentId:  types.StringValue(role.EnvironmentId),
+		CreatedAt:      types.StringValue(role.CreatedAt.String()),
+		UpdatedAt:      types.StringValue(role.UpdatedAt.String()),
+		Key:            types.StringValue(role.Key),
+		Name:           types.StringValue(role.Name),
+		Description:    types.StringPointerValue(role.Description),
+		Permissions:    data.Permissions,
+		Extends:        data.Extends,
+	}
+	return state, nil, diags
+}
+
+func compareSlicesValues(a []string, b []string) bool {
+	if len(a) != len(b) {
+		return false
+	}
+	for _, aVal := range a {
+		found := false
+		for _, bVal := range b {
+			if aVal == bVal {
+				found = true
+				break
+			}
+		}
+		if !found {
+			return false
+		}
+	}
+	return true
+}
+func (r *RoleClient) RoleCreate(ctx context.Context, rolePlan *RoleModel) (error, diag.Diagnostics) {
```

Comment:
pretty sure there is a missing new line here between the functions
execute go fmt